packages/scripts/src/monitoring-dvm2.0/voterStakes.ts

```diff
@@ -0,0 +1,59 @@
+// This script verifies the correct functioning of the VotingV2 contract. In particular, it verifies that, once all the
+// voters are processed by the updateTrackers function, the balance of the VotingV2 contract in VotingTokens is equal to
+// the sum of all users' stakes and pendingUnstakes. Otherwise it throws an error.
+// It can be run on a mainnet or goerli fork, with the following command and environment variables:
+// Or a goerli fork with:
+// CUSTOM_NODE_URL=https://<goerli OR mainnet>.infura.io/v3/<YOUR-INFURA-KEY> \
+// yarn hardhat run ./src/monitoring-dvm2.0/voterStakes.ts
+
+const hre = require("hardhat");
+import { VotingTokenEthers, VotingV2Ethers } from "@uma/contracts-node";
+import { getContractInstance } from "../utils/contracts";
+import { forkNetwork, getForkChainId } from "../utils/utils";
+import { getUniqueVoters, updateTrackers } from "./common";
+const { ethers } = hre;
+
+async function main() {
+  if (!process.env.CUSTOM_NODE_URL) throw new Error("CUSTOM_NODE_URL must be defined in env");
+  await forkNetwork(process.env.CUSTOM_NODE_URL);
+  const chainId = await getForkChainId(process.env.CUSTOM_NODE_URL);
+
+  if (!chainId || (chainId != 1 && chainId != 5)) throw new Error("This script should be run on mainnet or goerli");
+
+  const votingV2 = await getContractInstance<VotingV2Ethers>("VotingV2", undefined, chainId);
+  const votingToken = await getContractInstance<VotingTokenEthers>("VotingToken", undefined, chainId);
+
+  const uniqueVoters = await getUniqueVoters(votingV2);
+
+  // Update trackers for all voters
+  await updateTrackers(votingV2, uniqueVoters);
+
+  // get voter stakes for each voter
+  const voterStakes = await Promise.all(
+    uniqueVoters.map(async (voter) => {
+      const stake = await votingV2.voterStakes(voter);
+      return stake.stake.add(stake.pendingUnstake);
+    })
+  );
+
+  const sumStakes = voterStakes.reduce((a, b) => a.add(b), ethers.BigNumber.from(0));
+
+  const votingV2Balance = await votingToken.balanceOf(votingV2.address);
+
+  if (!sumStakes.eq(votingV2Balance)) {
```

Comment:
This would fail if someone had mistakenly transferred tokens to the contract

Reply:
that might be the point...this would show us if someone was to do this.

Reply:
Agree with both of you.

Reply:
Is it possible that this also breaks due to (expected) rounding down in our math?

Let's say we have three stakers a, b, and c:
1. They all have 100 wei staked.
2. c gets slashed for 9 wei. a and b are both correct.
3. 4 wei gets added to a's account and 4 wei gets added to b's account. 1 wei is lost.
4. This test breaks.

Do y'all agree? Are there more robust ways to create this check? I was thinking of a check with the following logic:
1. balanceOf(contract) >= sum of stakes. There should never be more tokens in the accounting than in the contract. That would always be bad.
2. balanceOf(contract) <= sum of stakes + 1e-9 UMA (i.e. 1 billion wei, which would be a lot of rounding error, since it should only be 1e-18 per rounding event).
3. If we wanted to handle accidental transfers, we could search for incoming transfers who called .transfer directly on the UMA token contract (we could just look for incoming transfers whose target was the UMA token contract), which would likely handle the vast majority of these. Then we could just add this to the sum of stakes so it's accounted for.

Reply:
Yes I agree with you @mrice32

I've added some changes to consider the offset due to the rounding in the checks + remove the votingToken balance coming from transfers.